clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -0,0 +1,1005 @@
+    }
+
+    /**
+     * Pretty-print an exception.
+     *
+     * @param throwable                         The exception.
+     *
+     * @return                                  A compact human-readable string.
+     */
+    static String prettyPrintException(Throwable throwable) {
+        if (throwable == null)
+            return "Null exception.";
+        if (throwable.getMessage() != null) {
+            return throwable.getMessage();
+        }
+        return throwable.getClass().getSimpleName();
+    }
+
+    /**
+     * Close an AutoCloseable, suppressing any exceptions.
+     *
+     * @param what                              Human-readable string explaining what we are closing.
+     * @param closeable                         What we are closing, or null to ignore.
+     */
+    static void closeQuietly(String what, AutoCloseable closeable) {
+        if (closeable == null)
+            return;
+        try {
+            closeable.close();
+        } catch (Throwable exc) {
+            log.error("Error closing " + what, exc);
+        }
+    }
+
+    static KafkaAdminClient create(AdminClientConfig config) {
+        Metadata metadata = null;
+        Metrics metrics = null;
+        Selector selector = null;
+        NetworkClient networkClient = null;
+        Time time = Time.SYSTEM;
+        String clientId = generateClientId(config);
+        ChannelBuilder channelBuilder = null;
+        ApiVersions apiVersions = new ApiVersions();
+
+        try {
+            metadata = new Metadata(config.getLong(AdminClientConfig.RETRY_BACKOFF_MS_CONFIG),
+                    config.getLong(AdminClientConfig.METADATA_MAX_AGE_CONFIG));
+            metrics = new Metrics(new MetricConfig(), new LinkedList<MetricsReporter>(), time);
+            String metricGrpPrefix = "adminclient";
+            channelBuilder = ClientUtils.createChannelBuilder(config);
+            selector = new Selector(config.getLong(AdminClientConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG),
+                    metrics, time, metricGrpPrefix, channelBuilder);
+            networkClient = new NetworkClient(selector,
+                metadata,
+                clientId,
+                100,
+                config.getLong(AdminClientConfig.RECONNECT_BACKOFF_MS_CONFIG),
+                config.getInt(AdminClientConfig.SEND_BUFFER_CONFIG),
+                config.getInt(AdminClientConfig.RECEIVE_BUFFER_CONFIG),
+                config.getInt(AdminClientConfig.REQUEST_TIMEOUT_MS_CONFIG),
+                time,
+                true,
+                apiVersions);
+            channelBuilder = null;
+            return new KafkaAdminClient(config, clientId, time, metadata, metrics, networkClient, selector);
+        } catch (Throwable exc) {
+            closeQuietly("Metrics", metrics);
+            closeQuietly("ChannelBuilder", channelBuilder);
+            closeQuietly("Selector", selector);
+            closeQuietly("NetworkClient", networkClient);
```

Comment:
I didn't realize that NetworkClient closed the Selector... but I looked at the code again, and indeed it does.  So we don't need to worry about that in `KafkaAdminClient#close`.  However, we do need to handle the case where the various constructors throw exceptions (pretty much all of them can.)  For example, the Selector constructor could fail and leave our ChannelBuilder dangling, and etc.